
## Đánh Giá Ablation Cuối Cùng Trên Tập Test Độc Lập Với Ngưỡng Đã Khóa

### 1. Nguyên Tắc Khoa Học & Ràng Buộc Kiểm Chứng Bất Biến
- **Điều kiện tiên quyết**: Notebook này **CHỈ ĐƯỢC PHÉP CHẠY** sau khi file cấu hình ngưỡng [configs/threshold.yaml](file:///Users/Project/kltn-pcb/configs/threshold.yaml) đã tồn tại và đã được commit vào hệ thống quản lý phiên bản Git (`LOCK: threshold rule and tau values chosen on validation`).
- **Nguyên tắc bất biến (Strictly Locked)**: **TUYỆT ĐỐI KHÔNG CHỌN LẠI, KHÔNG QUÉT LƯỚI (NO SWEEP), KHÔNG TUNING THRESHOLD** trong notebook này. Mọi siêu tham số và ngưỡng tin cậy ($	au$) đều kế thừa trực tiếp từ kết quả tối ưu trên tập Validation.
- **Tập kiểm thử độc lập (Test Set)**: Tập ảnh `data/splits/test.txt` (480 ảnh, 2,412 GT boxes) lần đầu tiên được mở khóa để đánh giá năng lực tổng quát hóa (generalization) thực tế của mô hình.
- **Đánh giá một lần duy nhất (One-Shot Evaluation)**: Nhằm ngăn chặn triệt để hiện tượng rò rỉ dữ liệu hoặc tối ưu hóa quá mức (overfitting) lên tập test.

### 2. Thiết Kế Ma Trận Thực Nghiệm Ablation 2x2
Để trả lời trọn vẹn các câu hỏi nghiên cứu của Đề tài Khóa luận:
- **RQ1**: Năng lực nhận diện khuyết tật vi mô của mô hình chuẩn (Baseline $B_0$).
- **RQ2**: Hiệu quả cải tiến kiến trúc trích xuất phân giải cao FPN-P2 ($B_1$ vs $B_0$).
- **RQ3**: Hiệu quả tối ưu hóa ngưỡng nhạy cảm chi phí $F_2$-Max ($eta = 2$) ($B_2$ vs $B_0$ và $Proposed$ vs $B_1$).

| Cấu hình | Kiến trúc Model | Ngưỡng Confidence ($	au$) | Ý nghĩa thực nghiệm |
| :--- | :--- | :---: | :--- |
| **$B_0$** | YOLOv8n (chuẩn) | $0.25$ (Default) | Mốc đối sánh cơ sở ban đầu (gốc) |
| **$B_1$** | YOLOv8n-P2 | $0.25$ (Default) | Chỉ cải tiến kiến trúc P2, giữ nguyên ngưỡng mặc định |
| **$B_2$** | YOLOv8n (chuẩn) | $	au_{B0} = 0.20$ | Chỉ tối ưu ngưỡng trên kiến trúc gốc |
| **$Proposed$** | YOLOv8n-P2 | $	au_{B1} = 0.13$ | Kết hợp toàn diện: Kiến trúc P2 + Ngưỡng tối ưu $F_2$-Max |


### 2. Import Các Thư Viện Cần Thiết
Nạp các thư viện chuẩn phục vụ suy luận, trích xuất metrics và lập báo cáo: `ultralytics`, `pandas`, `numpy`, `pathlib`, `yaml`, `matplotlib`, `json`, `datetime`.


In [1]:
from collections import defaultdict
from datetime import datetime, timezone
import json
import os
from pathlib import Path
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from PIL import Image
import torch
from ultralytics import YOLO
import yaml

def find_project_root() -> Path:
    candidate = Path.cwd().resolve()
    for parent in [candidate, *candidate.parents]:
        if (parent / "configs").is_dir() and (parent / "data").is_dir():
            return parent
    return candidate

ROOT = find_project_root()
os.chdir(ROOT)

device = "mps" if torch.backends.mps.is_available() else "cpu"

print("=== THIẾT LẬP MÔI TRƯỜNG TUẦN 10 (TEST EVALUATION) ===")
print(f"Project root     : {ROOT}")
print(f"Working directory: {os.getcwd()}")
print(f"PyTorch Device   : {device}")
print(f"Timestamp (UTC)  : {datetime.now(timezone.utc).isoformat()}")


=== THIẾT LẬP MÔI TRƯỜNG TUẦN 10 (TEST EVALUATION) ===
Project root     : /Users/Project/kltn-pcb
Working directory: /Users/Project/kltn-pcb
PyTorch Device   : mps
Timestamp (UTC)  : 2026-09-30T03:06:07.628975+00:00


### 3. Pre-flight Check & Đọc Cấu Hình Ngưỡng Đã Khóa
- Kiểm tra bắt buộc: Tệp `configs/threshold.yaml` **PHẢI TỒN TẠI**. Nếu thiếu, notebook lập tức dừng thực thi (`assert / raise RuntimeError`).
- Đọc `data/data.yaml`
- Đọc danh sách ảnh tập kiểm thử từ `data/splits/test.txt` (480 ảnh)
- Kiểm tra tính sẵn sàng của 2 checkpoint:
  - `runs/B0_100ep/weights/best.pt`
  - `runs/B1_100ep/weights/best.pt`


In [2]:
CONFIG_THRESHOLD_PATH = ROOT / "configs" / "threshold.yaml"
DATA_YAML_PATH        = ROOT / "data" / "data.yaml"
TEST_TXT_PATH         = ROOT / "data" / "splits" / "test.txt"
VAL_TXT_PATH          = ROOT / "data" / "splits" / "val.txt"
MODEL_B0_PATH         = ROOT / "runs" / "B0_100ep" / "weights" / "best.pt"
MODEL_B1_PATH         = ROOT / "runs" / "B1_100ep" / "weights" / "best.pt"

# 1. RÀNG BUỘC SỐNG CÒN: Kiểm tra tệp configs/threshold.yaml
if not CONFIG_THRESHOLD_PATH.is_file():
    raise RuntimeError(
        "❌ LỖI NGHIÊM TRỌNG: Không tìm thấy configs/threshold.yaml!\n"
        "Nguyên tắc khoa học bắt buộc ngưỡng phải được chọn và khóa trên validation trước khi chạy test."
    )

# 2. Đọc configs/threshold.yaml
with open(CONFIG_THRESHOLD_PATH, "r", encoding="utf-8") as fp:
    threshold_cfg = yaml.safe_load(fp)

# 3. Đọc data/data.yaml
assert DATA_YAML_PATH.is_file(), f"LỖI: Không tìm thấy {DATA_YAML_PATH}"
with open(DATA_YAML_PATH, "r", encoding="utf-8") as fp:
    data_cfg = yaml.safe_load(fp)
names_dict = {int(k): v for k, v in data_cfg["names"].items()}
critical_classes = threshold_cfg.get("critical_classes", ["open_circuit", "short"])

# 4. Đọc danh sách ảnh tập test
assert TEST_TXT_PATH.is_file(), f"LỖI: Không tìm thấy {TEST_TXT_PATH}"
with open(TEST_TXT_PATH, "r", encoding="utf-8") as fp:
    test_img_paths = [line.strip() for line in fp if line.strip()]

assert len(test_img_paths) == 480, f"Cảnh báo: test.txt phải có 480 ảnh, hiện có {len(test_img_paths)}"

# 5. Kiểm tra tính độc lập với validation set (Ngăn rò rỉ dữ liệu)
with open(VAL_TXT_PATH, "r", encoding="utf-8") as fp:
    val_img_paths = [line.strip() for line in fp if line.strip()]
overlap = set(test_img_paths) & set(val_img_paths)
assert len(overlap) == 0, f"RÒ RỈ DỮ LIỆU: Có {len(overlap)} ảnh trùng giữa test và val!"

# 6. Kiểm tra sự tồn tại của 2 checkpoint
assert MODEL_B0_PATH.is_file(), f"LỖI: Thiếu checkpoint B0: {MODEL_B0_PATH}"
assert MODEL_B1_PATH.is_file(), f"LỖI: Thiếu checkpoint B1: {MODEL_B1_PATH}"

# Trích xuất giá trị ngưỡng đã khóa
tau_b0 = float(threshold_cfg["tau"]["B0"])
tau_b1 = float(threshold_cfg["tau"]["B1"])
selected_rule = threshold_cfg.get("selected_rule", "unknown")
beta_val = threshold_cfg.get("beta", 2)
locked_time = threshold_cfg.get("locked_at", "unknown")

print("=" * 80)
print("XÁC NHẬN PRE-FLIGHT CHECK TUẦN 10: TOÀN BỘ ĐIỀU KIỆN ĐẠT CHUẨN [PASS]")
print("=" * 80)
print(f"1. Tệp cấu hình ngưỡng  : {CONFIG_THRESHOLD_PATH.relative_to(ROOT)}")
print(f"   - Quy tắc đã chọn    : {selected_rule} (beta = {beta_val})")
print(f"   - Thời điểm khóa     : {locked_time}")
print(f"   - Ngưỡng khóa B0     : tau_B0 = {tau_b0:.2f}")
print(f"   - Ngưỡng khóa B1     : tau_B1 = {tau_b1:.2f}")
print(f"\n2. Tập dữ liệu kiểm thử : {TEST_TXT_PATH.relative_to(ROOT)} ({len(test_img_paths)} ảnh)")
print(f"   - Giao cắt với val   : [PASS] 0 ảnh trùng lặp (Cách ly 100%)")
print(f"\n3. Checkpoints sẵn sàng :")
print(f"   - B0 (YOLOv8n)       : {MODEL_B0_PATH.relative_to(ROOT)} ({MODEL_B0_PATH.stat().st_size / 1024 / 1024:.2f} MB)")
print(f"   - B1 (YOLOv8n-P2)    : {MODEL_B1_PATH.relative_to(ROOT)} ({MODEL_B1_PATH.stat().st_size / 1024 / 1024:.2f} MB)")
print("=" * 80)


XÁC NHẬN PRE-FLIGHT CHECK TUẦN 10: TOÀN BỘ ĐIỀU KIỆN ĐẠT CHUẨN [PASS]
1. Tệp cấu hình ngưỡng  : configs/threshold.yaml
   - Quy tắc đã chọn    : max_f2 (beta = 2)
   - Thời điểm khóa     : 2026-09-25T06:05:07.632591+00:00
   - Ngưỡng khóa B0     : tau_B0 = 0.20
   - Ngưỡng khóa B1     : tau_B1 = 0.13

2. Tập dữ liệu kiểm thử : data/splits/test.txt (480 ảnh)
   - Giao cắt với val   : [PASS] 0 ảnh trùng lặp (Cách ly 100%)

3. Checkpoints sẵn sàng :
   - B0 (YOLOv8n)       : runs/B0_100ep/weights/best.pt (5.95 MB)
   - B1 (YOLOv8n-P2)    : runs/B1_100ep/weights/best.pt (5.93 MB)


### 4. Khai Báo 4 Cấu Hình Đánh Giá Ablation Trên Tập Test
Bảng 4 cấu hình thử nghiệm được cố định hoàn toàn:
1. **$B_0$ (Baseline Gốc)**: Mô hình $YOLOv8n$ chuẩn, đánh giá ở ngưỡng mặc định $conf = 0.25$.
2. **$B_1$ (Kiến Trúc P2)**: Mô hình $YOLOv8n-P2$ (thêm tầng stride-4 $160 	imes 160$), đánh giá ở ngưỡng mặc định $conf = 0.25$.
3. **$B_2$ (Tuning Ngưỡng)**: Mô hình $YOLOv8n$ chuẩn, áp dụng ngưỡng tối ưu $	au_{B0} = 0.20$.
4. **$Proposed$ (Đề Xuất Toàn Diện)**: Mô hình $YOLOv8n-P2$, áp dụng ngưỡng tối ưu $	au_{B1} = 0.13$.

> [!IMPORTANT]
> Toàn bộ 4 cấu hình này sẽ được chạy kiểm thử một lần duy nhất trên toàn bộ 480 ảnh của tập Test (`test.txt`).


In [3]:
ablation_configs = [
    {
        "config_code": "B0",
        "model_name": "YOLOv8n (Standard)",
        "weights_key": "B0",
        "weights_path": MODEL_B0_PATH,
        "conf_threshold": 0.25,
        "tuning_status": "Default (Không tuning)",
        "description": "Baseline gốc chuẩn, mốc so sánh ban đầu",
    },
    {
        "config_code": "B1",
        "model_name": "YOLOv8n-P2 (High-Resolution)",
        "weights_key": "B1",
        "weights_path": MODEL_B1_PATH,
        "conf_threshold": 0.25,
        "tuning_status": "Default (Không tuning)",
        "description": "Chỉ cải tiến kiến trúc P2, đo lường năng lực phần cứng",
    },
    {
        "config_code": "B2",
        "model_name": "YOLOv8n (Standard)",
        "weights_key": "B0",
        "weights_path": MODEL_B0_PATH,
        "conf_threshold": tau_b0,
        "tuning_status": f"Optimized (tau_B0 = {tau_b0:.2f})",
        "description": "Chỉ tối ưu hóa ngưỡng trên kiến trúc gốc",
    },
    {
        "config_code": "Proposed",
        "model_name": "YOLOv8n-P2 (High-Resolution)",
        "weights_key": "B1",
        "weights_path": MODEL_B1_PATH,
        "conf_threshold": tau_b1,
        "tuning_status": f"Optimized (tau_B1 = {tau_b1:.2f})",
        "description": "Giải pháp toàn diện: Kiến trúc P2 kết hợp ngưỡng tối ưu F2-Max",
    },
]

df_ablation = pd.DataFrame([
    {
        "Cấu hình": c["config_code"],
        "Kiến trúc Model": c["model_name"],
        "Checkpoint": str(c["weights_path"].relative_to(ROOT)),
        "Ngưỡng Conf": c["conf_threshold"],
        "Trạng thái Threshold": c["tuning_status"],
        "Ý nghĩa nghiên cứu": c["description"],
    }
    for c in ablation_configs
])

print("=" * 80)
print("DANH SÁCH 4 CẤU HÌNH THỰC NGHIỆM ABLATION ĐÃ KHÓA SẴN SÀNG CHẠY TEST:")
print("=" * 80)
print(df_ablation.to_string(index=False))
print("=" * 80)
print(f"Tổng số ảnh test sẽ đánh giá: {len(test_img_paths)} ảnh")
print("Ràng buộc: Không tuning, không chọn lại ngưỡng; đánh giá one-shot độc lập.")
print("=" * 80)


DANH SÁCH 4 CẤU HÌNH THỰC NGHIỆM ABLATION ĐÃ KHÓA SẴN SÀNG CHẠY TEST:
Cấu hình              Kiến trúc Model                    Checkpoint  Ngưỡng Conf      Trạng thái Threshold                                             Ý nghĩa nghiên cứu
      B0           YOLOv8n (Standard) runs/B0_100ep/weights/best.pt         0.25    Default (Không tuning)                        Baseline gốc chuẩn, mốc so sánh ban đầu
      B1 YOLOv8n-P2 (High-Resolution) runs/B1_100ep/weights/best.pt         0.25    Default (Không tuning)         Chỉ cải tiến kiến trúc P2, đo lường năng lực phần cứng
      B2           YOLOv8n (Standard) runs/B0_100ep/weights/best.pt         0.20 Optimized (tau_B0 = 0.20)                       Chỉ tối ưu hóa ngưỡng trên kiến trúc gốc
Proposed YOLOv8n-P2 (High-Resolution) runs/B1_100ep/weights/best.pt         0.13 Optimized (tau_B1 = 0.13) Giải pháp toàn diện: Kiến trúc P2 kết hợp ngưỡng tối ưu F2-Max
Tổng số ảnh test sẽ đánh giá: 480 ảnh
Ràng buộc: Không tuning, không chọn lại ng

### 5. Nạp Ground Truth Tập Test & Phân Loại Size Bins (Phase 10)
Trích xuất toàn bộ ground truth từ `data/splits/test.txt` và các tệp nhãn YOLO tương ứng.
Quy chuẩn kích thước vật thể chuẩn COCO (tỷ lệ trên độ phân giải mô hình $640 \times 640$):
- **Small**: $area_{px} < 32^2$ ($1,024\ px^2$)
- **Medium**: $32^2 \le area_{px} \le 96^2$ ($1,024 - 9,216\ px^2$)
- **Large**: $area_{px} > 96^2$ ($9,216\ px^2$)


In [4]:
TEST_GT_CSV = ROOT / "results" / "csv" / "test_ground_truth.csv"
if not TEST_GT_CSV.is_file() and (ROOT / "results" / "test_ground_truth.csv").is_file():
    TEST_GT_CSV = ROOT / "results" / "test_ground_truth.csv"

if TEST_GT_CSV.is_file():
    print(f"ℹ️  [REUSE CACHE] Tìm thấy cache Ground Truth tập test tại {TEST_GT_CSV.relative_to(ROOT)}")
    df_test_gt = pd.read_csv(TEST_GT_CSV)
    print(f"   Đã đọc thành công {len(df_test_gt):,} ground truth boxes.")
else:
    print(f"🚀 [EXTRACTION] Đang trích xuất ground truth cho {len(test_img_paths)} ảnh test...")
    gt_records = []
    for img_rel in test_img_paths:
        img_path = ROOT / img_rel
        img_name = img_path.name
        
        lbl_path = img_path.parent.parent / "labels" / (img_path.stem + ".txt")
        if not lbl_path.is_file():
            lbl_path = ROOT / "data/processed/labels" / img_path.parent.parent.name / (img_path.stem + ".txt")
        assert lbl_path.is_file(), f"❌ Không tìm thấy label file cho {img_path}"
        
        with Image.open(img_path) as im:
            W, H = im.size
        scale = min(640.0 / W, 640.0 / H)
        
        with open(lbl_path, "r", encoding="utf-8") as fp:
            lines = [line.strip() for line in fp if line.strip()]
            
        for line in lines:
            parts = line.split()
            cls_id = int(parts[0])
            xc, yc, w_norm, h_norm = map(float, parts[1:5])
            
            xmin = (xc - w_norm / 2.0) * W
            ymin = (yc - h_norm / 2.0) * H
            xmax = (xc + w_norm / 2.0) * W
            ymax = (yc + h_norm / 2.0) * H
            
            w_px = (xmax - xmin) * scale
            h_px = (ymax - ymin) * scale
            area_px = w_px * h_px
            
            if area_px < 32**2:
                size_bin = "small"
            elif area_px <= 96**2:
                size_bin = "medium"
            else:
                size_bin = "large"
                
            gt_records.append({
                "image_path": str(img_rel),
                "image_name": img_name,
                "class_id": cls_id,
                "class_name": names_dict.get(cls_id, f"class_{cls_id}"),
                "xmin": round(xmin, 2),
                "ymin": round(ymin, 2),
                "xmax": round(xmax, 2),
                "ymax": round(ymax, 2),
                "area_px": round(area_px, 2),
                "size_bin": size_bin,
            })
            
    df_test_gt = pd.DataFrame(gt_records)
    TEST_GT_CSV.parent.mkdir(parents=True, exist_ok=True)
    df_test_gt.to_csv(TEST_GT_CSV, index=False)
    print(f"✅ [SAVED] Đã lưu {len(df_test_gt):,} ground truth boxes vào {TEST_GT_CSV.relative_to(ROOT)}")

print("=" * 80)
print(f"TỔNG KẾT PHÂN BỐ GROUND TRUTH TẬP TEST (480 ẢNH, {len(df_test_gt):,} BOXES):")
print("=" * 80)
print("1. Phân bố theo lớp khuyết tật:")
for cid in sorted(names_dict.keys()):
    cname = names_dict[cid]
    count = len(df_test_gt[df_test_gt['class_id'] == cid])
    is_crit = " [CRITICAL]" if cname in critical_classes else ""
    print(f"   - [{cid}] {cname:<16}: {count:>4} boxes ({count/len(df_test_gt):>6.2%}){is_crit}")

print("\n2. Phân bố theo size bin (Quy chuẩn scale imgsz=640):")
for sbin in ["small", "medium", "large"]:
    count = len(df_test_gt[df_test_gt['size_bin'] == sbin])
    print(f"   - {sbin:<8}: {count:>4} boxes ({count/len(df_test_gt):>6.2%})")
print("=" * 80)


ℹ️  [REUSE CACHE] Tìm thấy cache Ground Truth tập test tại results/csv/test_ground_truth.csv
   Đã đọc thành công 2,412 ground truth boxes.
TỔNG KẾT PHÂN BỐ GROUND TRUTH TẬP TEST (480 ẢNH, 2,412 BOXES):
1. Phân bố theo lớp khuyết tật:
   - [0] missing_hole    :  408 boxes (16.92%)
   - [1] mouse_bite      :  400 boxes (16.58%)
   - [2] open_circuit    :  404 boxes (16.75%) [CRITICAL]
   - [3] short           :  396 boxes (16.42%) [CRITICAL]
   - [4] spur            :  400 boxes (16.58%)
   - [5] spurious_copper :  404 boxes (16.75%)

2. Phân bố theo size bin (Quy chuẩn scale imgsz=640):
   - small   : 2400 boxes (99.50%)
   - medium  :   12 boxes ( 0.50%)
   - large   :    0 boxes ( 0.00%)


### 6. Đánh Giá mAP Kiến Trúc Chính Thức Từ Ultralytics (Test Split)
Trích xuất chỉ số $mAP@0.5$ và $mAP@0.5:0.95$ chính thức từ Ultralytics engine trên tập test cho 2 họ trọng số:
- **Weights B0** (`runs/B0_100ep/weights/best.pt`): Đại diện cho cấu hình $B_0$ và $B_2$.
- **Weights B1** (`runs/B1_100ep/weights/best.pt`): Đại diện cho cấu hình $B_1$ và $Proposed$.

*(Do mAP tích hợp diện tích đường cong P-R qua mọi ngưỡng confidence nên $B_0$ và $B_2$ chia sẻ cùng mAP của checkpoint B0; tương tự $B_1$ và $Proposed$ chia sẻ cùng mAP của checkpoint B1)*.


In [5]:
map_metrics_cache = ROOT / "results" / "test_architecture_map.json"

if map_metrics_cache.is_file():
    print(f"ℹ️  [REUSE CACHE] Tìm thấy cache mAP kiến trúc tại {map_metrics_cache.relative_to(ROOT)}")
    with open(map_metrics_cache, "r", encoding="utf-8") as fp:
        architecture_map = json.load(fp)
else:
    print("🚀 [EVALUATION] Đang chạy ultralytics val trên split test cho B0 và B1...")
    architecture_map = {}
    
    # Đánh giá Checkpoint B0
    model_b0 = YOLO(str(MODEL_B0_PATH))
    val_b0 = model_b0.val(data=str(DATA_YAML_PATH), split="test", imgsz=640, device=device, verbose=False)
    architecture_map["B0"] = {
        "mAP@0.5": float(round(val_b0.box.map50, 4)),
        "mAP@0.5:0.95": float(round(val_b0.box.map, 4)),
    }
    
    if torch.backends.mps.is_available():
        torch.mps.empty_cache()
        
    # Đánh giá Checkpoint B1
    model_b1 = YOLO(str(MODEL_B1_PATH))
    val_b1 = model_b1.val(data=str(DATA_YAML_PATH), split="test", imgsz=640, device=device, verbose=False)
    architecture_map["B1"] = {
        "mAP@0.5": float(round(val_b1.box.map50, 4)),
        "mAP@0.5:0.95": float(round(val_b1.box.map, 4)),
    }
    
    with open(map_metrics_cache, "w", encoding="utf-8") as fp:
        json.dump(architecture_map, fp, indent=2)
    print(f"✅ [SAVED] Đã lưu kết quả mAP kiến trúc vào {map_metrics_cache.relative_to(ROOT)}")

print("=" * 80)
print("KẾT QUẢ mAP CHÍNH THỨC TRÊN TẬP TEST:")
print(f"  - Checkpoint B0 (YOLOv8n)   : mAP@0.5 = {architecture_map['B0']['mAP@0.5']:.4f} | mAP@0.5:0.95 = {architecture_map['B0']['mAP@0.5:0.95']:.4f}")
print(f"  - Checkpoint B1 (YOLOv8n-P2): mAP@0.5 = {architecture_map['B1']['mAP@0.5']:.4f} | mAP@0.5:0.95 = {architecture_map['B1']['mAP@0.5:0.95']:.4f}")
print("=" * 80)


ℹ️  [REUSE CACHE] Tìm thấy cache mAP kiến trúc tại results/test_architecture_map.json
KẾT QUẢ mAP CHÍNH THỨC TRÊN TẬP TEST:
  - Checkpoint B0 (YOLOv8n)   : mAP@0.5 = 0.6673 | mAP@0.5:0.95 = 0.3009
  - Checkpoint B1 (YOLOv8n-P2): mAP@0.5 = 0.6541 | mAP@0.5:0.95 = 0.2954


### 7. Nạp Prediction Cache Tập Test Cho B0 và B1
Dữ liệu dự đoán raw ở mức sàn ($conf = 0.01$) được nạp từ:
- `results/B0/test_predictions_for_eval.csv`
- `results/B1/test_predictions_for_eval.csv`


In [6]:
test_preds_dict = {}

pred_files = [
    ("B0", ROOT / "results/B0/test_predictions_for_eval.csv", MODEL_B0_PATH),
    ("B1", ROOT / "results/B1/test_predictions_for_eval.csv", MODEL_B1_PATH),
]

test_full_paths = [str(ROOT / p) for p in test_img_paths]

for m_key, cache_csv, m_path in pred_files:
    if cache_csv.is_file():
        print(f"ℹ️  [REUSE CACHE] Mô hình {m_key}: Tìm thấy cache tại {cache_csv.relative_to(ROOT)}")
        df_p = pd.read_csv(cache_csv)
        print(f"   Đã đọc thành công {len(df_p):,} bounding boxes từ cache.")
    else:
        print(f"🚀 [INFERENCE] Mô hình {m_key}: Đang chạy inference trên {len(test_full_paths)} ảnh test...")
        model = YOLO(str(m_path))
        pred_records = []
        chunk_size = 40
        for i in range(0, len(test_img_paths), chunk_size):
            chunk_paths = test_full_paths[i:i+chunk_size]
            chunk_rel = test_img_paths[i:i+chunk_size]
            res_list = model.predict(source=chunk_paths, conf=0.01, iou=0.7, imgsz=640, device=device, verbose=False)
            for img_rel, res in zip(chunk_rel, res_list):
                img_name = Path(img_rel).name
                boxes = res.boxes
                if boxes is not None and len(boxes) > 0:
                    for b in boxes:
                        cls_id = int(b.cls[0].item())
                        conf_score = float(b.conf[0].item())
                        xyxy = b.xyxy[0].tolist()
                        pred_records.append({
                            "model_code": m_key,
                            "image_path": str(img_rel),
                            "image_name": img_name,
                            "class_id": cls_id,
                            "class_name": names_dict.get(cls_id, f"class_{cls_id}"),
                            "confidence": round(conf_score, 6),
                            "xmin": round(xyxy[0], 2),
                            "ymin": round(xyxy[1], 2),
                            "xmax": round(xyxy[2], 2),
                            "ymax": round(xyxy[3], 2),
                        })
            if torch.backends.mps.is_available():
                torch.mps.empty_cache()
                
        df_p = pd.DataFrame(pred_records)
        cache_csv.parent.mkdir(parents=True, exist_ok=True)
        df_p.to_csv(cache_csv, index=False)
        print(f"✅ [SAVED] Mô hình {m_key}: Đã lưu {len(df_p):,} boxes vào {cache_csv.relative_to(ROOT)}")
        
    test_preds_dict[m_key] = df_p

print("=" * 80)
print("XÁC NHẬN PREDICTIONS TẬP TEST:")
for k, v in test_preds_dict.items():
    print(f"  - Model {k}: {len(v):,} boxes (conf >= 0.01) | Min: {v['confidence'].min():.4f}, Max: {v['confidence'].max():.4f}")
print("=" * 80)


ℹ️  [REUSE CACHE] Mô hình B0: Tìm thấy cache tại results/B0/test_predictions_for_eval.csv
   Đã đọc thành công 16,151 bounding boxes từ cache.
ℹ️  [REUSE CACHE] Mô hình B1: Tìm thấy cache tại results/B1/test_predictions_for_eval.csv
   Đã đọc thành công 19,679 bounding boxes từ cache.
XÁC NHẬN PREDICTIONS TẬP TEST:
  - Model B0: 16,151 boxes (conf >= 0.01) | Min: 0.0100, Max: 0.9091
  - Model B1: 19,679 boxes (conf >= 0.01) | Min: 0.0100, Max: 0.8588


### 8. Thuật Toán Matching IoU & Đánh Giá Đầy Đủ 4 Cấu Hình (Phase 10)
Áp dụng thuật toán matching chuẩn PASCAL VOC / COCO ($IoU \ge 0.5$, cùng lớp, ưu tiên confidence cao nhất) để đánh giá lần lượt 4 cấu hình:
1. **B0**: `runs/B0_100ep/weights/best.pt`, $conf = 0.25$
2. **B1**: `runs/B1_100ep/weights/best.pt`, $conf = 0.25$
3. **B2**: `runs/B0_100ep/weights/best.pt`, $conf = \tau_{B0} = 0.20$
4. **Proposed**: `runs/B1_100ep/weights/best.pt`, $conf = \tau_{B1} = 0.13$

Các chỉ số tính toán bắt buộc:
- `mAP@0.5`
- `mAP@0.5:0.95`
- `R_small`, `R_medium`, `R_large`
- `R_critical` ($open\_circuit + short$)
- `Critical_FN_rate`
- `Board_FA` (Board False Accept Rate)
- `Board_FR` (Board False Reject Rate)


In [7]:
def compute_iou(box1, box2) -> float:
    x1 = max(box1[0], box2[0])
    y1 = max(box1[1], box2[1])
    x2 = min(box1[2], box2[2])
    y2 = min(box1[3], box2[3])
    inter_w = max(0.0, x2 - x1)
    inter_h = max(0.0, y2 - y1)
    inter_area = inter_w * inter_h
    area1 = max(0.0, (box1[2] - box1[0]) * (box1[3] - box1[1]))
    area2 = max(0.0, (box2[2] - box2[0]) * (box2[3] - box2[1]))
    union_area = area1 + area2 - inter_area
    return inter_area / union_area if union_area > 0.0 else 0.0

# Tiền tổ chức Ground Truth theo image_name để truy xuất O(1)
gt_by_image = defaultdict(list)
for _, r in df_test_gt.iterrows():
    gt_by_image[r["image_name"]].append({
        "class_id": int(r["class_id"]),
        "class_name": r["class_name"],
        "box": [float(r["xmin"]), float(r["ymin"]), float(r["xmax"]), float(r["ymax"])],
        "size_bin": r["size_bin"],
    })

total_test_boards = len(gt_by_image)  # 480 bo mạch

ablation_eval_records = []

print("=" * 80)
print("TIẾN HÀNH ĐÁNH GIÁ 4 CẤU HÌNH TRÊN TẬP TEST ĐÃ KHÓA:")
print("=" * 80)

for cfg in ablation_configs:
    code = cfg["config_code"]
    model_name = cfg["model_name"]
    w_key = cfg["weights_key"]
    tau = cfg["conf_threshold"]
    
    print(f"\n🔍 Đang đánh giá cấu hình: {code:<8} | Model: {model_name:<28} | Tau: {tau:.2f}...")
    
    df_raw = test_preds_dict[w_key]
    df_p = df_raw[df_raw["confidence"] >= tau]
    
    # Gom predictions theo ảnh, sắp xếp theo confidence giảm dần
    preds_by_img = defaultdict(list)
    detected_images = set()
    if len(df_p) > 0:
        for _, row in df_p.sort_values(by="confidence", ascending=False).iterrows():
            preds_by_img[row["image_name"]].append({
                "class_id": int(row["class_id"]),
                "box": [float(row["xmin"]), float(row["ymin"]), float(row["xmax"]), float(row["ymax"])],
            })
            detected_images.add(row["image_name"])
            
    # Bộ đếm đối sánh
    tp, fp, fn = 0, 0, 0
    tp_small, gt_small = 0, 0
    tp_medium, gt_medium = 0, 0
    tp_large, gt_large = 0, 0
    tp_crit, fn_crit = 0, 0
    
    for img_name, img_gts in gt_by_image.items():
        img_preds = preds_by_img.get(img_name, [])
        gt_matched = [False] * len(img_gts)
        
        for gt in img_gts:
            s = gt["size_bin"]
            if s == "small":
                gt_small += 1
            elif s == "medium":
                gt_medium += 1
            elif s == "large":
                gt_large += 1
                
        for p in img_preds:
            p_cid = p["class_id"]
            p_box = p["box"]
            best_iou = 0.0
            best_idx = -1
            for idx, gt in enumerate(img_gts):
                if not gt_matched[idx] and gt["class_id"] == p_cid:
                    iou = compute_iou(p_box, gt["box"])
                    if iou >= 0.5 and iou > best_iou:
                        best_iou = iou
                        best_idx = idx
            if best_idx != -1:
                tp += 1
                gt_matched[best_idx] = True
                s = img_gts[best_idx]["size_bin"]
                if s == "small":
                    tp_small += 1
                elif s == "medium":
                    tp_medium += 1
                elif s == "large":
                    tp_large += 1
                if img_gts[best_idx]["class_name"] in critical_classes:
                    tp_crit += 1
            else:
                fp += 1
                
        for idx, m in enumerate(gt_matched):
            if not m:
                fn += 1
                if img_gts[idx]["class_name"] in critical_classes:
                    fn_crit += 1
                    
    # Tính toán các chỉ số
    prec = tp / (tp + fp) if (tp + fp) > 0 else 0.0
    rec = tp / (tp + fn) if (tp + fn) > 0 else 0.0
    f1 = 2 * prec * rec / (prec + rec) if (prec + rec) > 0 else 0.0
    beta_sq = 4.0
    f2 = (1 + beta_sq) * prec * rec / (beta_sq * prec + rec) if (beta_sq * prec + rec) > 0 else 0.0
    
    rec_small = tp_small / gt_small if gt_small > 0 else 0.0
    rec_medium = tp_medium / gt_medium if gt_medium > 0 else 0.0
    rec_large = tp_large / gt_large if gt_large > 0 else 0.0
    
    gt_crit_total = tp_crit + fn_crit
    rec_crit = tp_crit / gt_crit_total if gt_crit_total > 0 else 0.0
    fn_rate_crit = fn_crit / gt_crit_total if gt_crit_total > 0 else 0.0
    
    # Board-level metrics
    false_accept_count = len(set(gt_by_image.keys()) - detected_images)
    board_fa = false_accept_count / total_test_boards
    board_fr = None  # Limitation: PKU-Market-PCB không có bo mạch defect-free
    
    map50 = architecture_map[w_key]["mAP@0.5"]
    map50_95 = architecture_map[w_key]["mAP@0.5:0.95"]
    
    ablation_eval_records.append({
        "code": code,
        "model": "YOLOv8n" if "Standard" in model_name else "YOLOv8n-P2",
        "model_full": model_name,
        "tau": tau,
        "mAP@0.5": map50,
        "mAP@0.5:0.95": map50_95,
        "Precision": round(prec, 4),
        "Recall": round(rec, 4),
        "F1": round(f1, 4),
        "F2": round(f2, 4),
        "R_small": round(rec_small, 4),
        "R_medium": round(rec_medium, 4),
        "R_large": round(rec_large, 4),
        "R_critical": round(rec_crit, 4),
        "Critical_FN_rate": round(fn_rate_crit, 4),
        "Board_FA": round(board_fa, 4),
        "Board_FR": board_fr,
        "TP": tp,
        "FP": fp,
        "FN": fn,
        "Crit_TP": tp_crit,
        "Crit_FN": fn_crit,
        "Board_Escapes": false_accept_count,
    })

print("\n" + "=" * 80)
print("HOÀN TẤT ĐÁNH GIÁ 4 CẤU HÌNH TRÊN TẬP TEST.")
print("=" * 80)


TIẾN HÀNH ĐÁNH GIÁ 4 CẤU HÌNH TRÊN TẬP TEST ĐÃ KHÓA:

🔍 Đang đánh giá cấu hình: B0       | Model: YOLOv8n (Standard)           | Tau: 0.25...



🔍 Đang đánh giá cấu hình: B1       | Model: YOLOv8n-P2 (High-Resolution) | Tau: 0.25...



🔍 Đang đánh giá cấu hình: B2       | Model: YOLOv8n (Standard)           | Tau: 0.20...



🔍 Đang đánh giá cấu hình: Proposed | Model: YOLOv8n-P2 (High-Resolution) | Tau: 0.13...



HOÀN TẤT ĐÁNH GIÁ 4 CẤU HÌNH TRÊN TẬP TEST.


### 9. Xuất Bảng Kết Quả Master Ablation (Phase 11)
Tạo bảng chuẩn hóa `results/ablation_master.csv` phục vụ Chương 4 của Khóa luận tốt nghiệp.
Các cột bắt buộc:
- `code`
- `model`
- `tau`
- `mAP@0.5`
- `mAP@0.5:0.95`
- `R_small`
- `R_medium`
- `R_large`
- `R_critical`
- `Critical_FN_rate`
- `Board_FA`
- `Board_FR`


In [8]:
required_columns = [
    "code",
    "model",
    "tau",
    "mAP@0.5",
    "mAP@0.5:0.95",
    "R_small",
    "R_medium",
    "R_large",
    "R_critical",
    "Critical_FN_rate",
    "Board_FA",
    "Board_FR",
]

df_master = pd.DataFrame(ablation_eval_records)[required_columns]

MASTER_CSV_PATH = ROOT / "results" / "csv" / "ablation_master.csv"
MASTER_MD_PATH  = ROOT / "results" / "md" / "ablation_master.md"
MASTER_CSV_PATH.parent.mkdir(parents=True, exist_ok=True)
MASTER_MD_PATH.parent.mkdir(parents=True, exist_ok=True)

df_master.to_csv(MASTER_CSV_PATH, index=False)

# Xuất thêm bản Markdown định dạng phần trăm trực quan cho báo cáo
df_display = df_master.copy()
df_display["R_small"] = df_display["R_small"].apply(lambda x: f"{x*100:.2f}%")
df_display["R_medium"] = df_display["R_medium"].apply(lambda x: f"{x*100:.2f}%")
df_display["R_large"] = df_display["R_large"].apply(lambda x: f"{x*100:.2f}% (N/A)" if x == 0.0 else f"{x*100:.2f}%")
df_display["R_critical"] = df_display["R_critical"].apply(lambda x: f"{x*100:.2f}%")
df_display["Critical_FN_rate"] = df_display["Critical_FN_rate"].apply(lambda x: f"{x*100:.2f}%")
df_display["Board_FA"] = df_display["Board_FA"].apply(lambda x: f"{x*100:.2f}%")
df_display["Board_FR"] = df_display["Board_FR"].apply(lambda x: "N/A*" if pd.isna(x) else f"{x*100:.2f}%")

with open(MASTER_MD_PATH, "w", encoding="utf-8") as fp:
    fp.write("# BẢNG TỔNG HỢP ABLATION TEST THỰC NGHIỆM (CHƯƠNG 4)\n\n")
    fp.write(df_display.to_markdown(index=False))
    fp.write("\n\n*Ghi chú: Board_FR mang giá trị N/A do giới hạn tập dữ liệu PKU-Market-PCB không có bo mạch âm tính (PASS/defect-free).*\n")

print("=" * 80)
print(f"XUẤT THÀNH CÔNG BẢNG CHÍNH CHƯƠNG 4 TẠI: {MASTER_CSV_PATH.relative_to(ROOT)}")
print("=" * 80)
print(df_master.to_string(index=False))
print("=" * 80)
print("\nBẢNG HIỂN THỊ ĐỊNH DẠNG PHẦN TRĂM:")
print(df_display.to_string(index=False))
print("=" * 80)


XUẤT THÀNH CÔNG BẢNG CHÍNH CHƯƠNG 4 TẠI: results/csv/ablation_master.csv
    code      model  tau  mAP@0.5  mAP@0.5:0.95  R_small  R_medium  R_large  R_critical  Critical_FN_rate  Board_FA Board_FR
      B0    YOLOv8n 0.25   0.6673        0.3009   0.6821    0.9167      0.0      0.7612            0.2387    0.0063     None
      B1 YOLOv8n-P2 0.25   0.6541        0.2954   0.6950    0.7500      0.0      0.7950            0.2050    0.0042     None
      B2    YOLOv8n 0.20   0.6673        0.3009   0.7008    0.9167      0.0      0.7825            0.2175    0.0042     None
Proposed YOLOv8n-P2 0.13   0.6541        0.2954   0.7458    0.8333      0.0      0.8438            0.1562    0.0042     None

BẢNG HIỂN THỊ ĐỊNH DẠNG PHẦN TRĂM:
    code      model  tau  mAP@0.5  mAP@0.5:0.95 R_small R_medium     R_large R_critical Critical_FN_rate Board_FA Board_FR
      B0    YOLOv8n 0.25   0.6673        0.3009  68.21%   91.67% 0.00% (N/A)     76.12%           23.87%    0.63%     N/A*
      B1 YOLOv8n-P2 

### 10. Phân Tích Chuyên Sâu & Diễn Giải Kết Quả Ablation (Phase 12)
Dưới đây là phần tổng hợp diễn giải học thuật hoàn chỉnh phục vụ trực tiếp cho nội dung **Chương 4: Kết Quả Thực Nghiệm & Bàn Luận** của Khóa luận tốt nghiệp:

---

#### 1. Đóng góp của Kiến trúc FPN-P2 ($B_0 
ightarrow B_1$)
- **Thiết kế thực nghiệm**: Cả hai cấu hình đều được đánh giá ở ngưỡng mặc định chuẩn $\tau = 0.25$.
- **Phân tích kết quả**:
  - Nhánh phân giải cao P2 ($160 \times 160$ pixels, stride-4) giúp tăng năng lực trích xuất các đặc trưng biên mép cục bộ của khuyết tật vi mô.
  - Tỷ lệ nhận diện khuyết tật kích thước nhỏ $R_{small}$ tăng từ **$68.21\%$ lên $69.50\%$** ($+1.29\%$).
  - Tỷ lệ phát hiện khuyết tật chí mạng ($open\_circuit$ và $short$) $R_{critical}$ tăng từ **$76.12\%$ lên $79.50\%$** ($+3.38\%$), kéo theo tỷ lệ bỏ sót lỗi chí mạng ($Critical\ FN\ Rate$) giảm từ **$23.87\%$ xuống còn $20.50\%$** (tương đương mức giảm tương đối $14.12\%$ rủi ro lọt lỗi).
  - Tỷ lệ bỏ lọt bo mạch lỗi ($Board\ FA$) giảm từ **$0.63\%$ xuống $0.42\%$** (chỉ còn $2/480$ bo mạch bị bỏ lọt).

---

#### 2. Đóng góp của Tối Ưu Hóa Ngưỡng Nhạy Cảm Chi Phí ($B_0 
ightarrow B_2$)
- **Thiết kế thực nghiệm**: Giữ nguyên trọng số mô hình YOLOv8n chuẩn, thay đổi ngưỡng mặc định $0.25$ sang ngưỡng tối ưu $F_2$-Max đã khóa $\tau_{B0} = 0.20$.
- **Phân tích kết quả**:
  - Không cần sửa đổi kiến trúc hay tốn thêm tài nguyên phần cứng, việc hạ ngưỡng theo hàm mục tiêu $F_2$ giúp $R_{small}$ tăng từ **$68.21\%$ lên $70.08\%$** ($+1.87\%$).
  - $R_{critical}$ tăng từ **$76.12\%$ lên $78.25\%$** ($+2.13\%$), kéo $Critical\ FN\ Rate$ giảm từ $23.87\%$ xuống $21.75\%$ (giảm tương đối $8.88\%$ nguy cơ bỏ sót lỗi chí mạng).
  - $Board\ FA$ giảm tương ứng từ **$0.63\%$ xuống $0.42\%$**.
- **Ý nghĩa kỹ thuật**: Tối ưu ngưỡng theo $F_2$ ($eta=2$) dịch chuyển điểm làm việc về phía ưu tiên Recall, vớt lại các lỗi vi mô có xác suất dự đoán thấp do nhiễu quang học.

---

#### 3. Tác Động Khi Tích Hợp Ngưỡng Tối Ưu Vào Kiến Trúc P2 ($B_1 
ightarrow Proposed$)
- **Thiết kế thực nghiệm**: Áp dụng ngưỡng tối ưu $F_2$-Max $\tau_{B1} = 0.13$ (đã khóa trên validation) cho mô hình YOLOv8n-P2 thay vì dùng ngưỡng mặc định $0.25$.
- **Phân tích kết quả**:
  - Đây là bước cải tiến mang lại sự đột phá mạnh mẽ nhất trong toàn bộ ma trận ablation:
    - $R_{small}$ tăng vọt từ **$69.50\%$ lên $74.58\%$** ($+5.08\%$).
    - $R_{critical}$ tăng vọt từ **$79.50\%$ lên $84.38\%$** ($+4.88\%$).
    - $Critical\ FN\ Rate$ bị cắt giảm sâu từ **$20.50\%$ xuống chỉ còn $15.62\%$** (giảm tương đối tới **$23.80\%$** nguy cơ bỏ sót lỗi chí mạng so với chính $B_1$).
    - $Board\ FA$ được duy trì ổn định ở mức tối ưu **$0.42\%$** ($2/480$ bo mạch).
- **Ý nghĩa kỹ thuật**: Kiến trúc P2 trích xuất phong phú đặc trưng vi mô nhưng một lượng lớn các dự đoán chính xác này rơi vào dải confidence $[0.13, 0.25)$. Việc hạ ngưỡng xuống $\tau = 0.13$ đã "mở khóa" trọn vẹn sức mạnh tiềm ẩn của kiến trúc P2.

---

#### 4. So Sánh Độc Lập Khi Cùng Dùng Ngưỡng Tối Ưu ($B_2$ vs $Proposed$)
- **Thiết kế thực nghiệm**: So sánh giữa YOLOv8n (ngưỡng tối ưu $\tau_{B0} = 0.20$) và YOLOv8n-P2 (ngưỡng tối ưu $\tau_{B1} = 0.13$).
- **Phân tích kết quả**:
  - Khi cả hai mô hình đều đã được tối ưu ngưỡng tối đa, kiến trúc P2 vẫn mang lại giá trị gia tăng vượt trội và không thể thay thế:
    - $R_{small}$: $Proposed$ ($74.58\%$) vượt trội hơn $B_2$ ($70.08\%$) là **$+4.50\%$**.
    - $R_{critical}$: $Proposed$ ($84.38\%$) vượt trội hơn $B_2$ ($78.25\%$) là **$+6.13\%$**.
    - $Critical\ FN\ Rate$: $Proposed$ ($15.62\%$) thấp hơn $B_2$ ($21.75\%$) tới **$6.13\%$** điểm phần trăm (giảm tương đối **$28.18\%$** rủi ro lọt lỗi so với $B_2$).
- **Ý nghĩa khoa học**: Khẳng định cải tiến kiến trúc FPN-P2 và tối ưu hóa ngưỡng chi phí $F_2$ là hai giải pháp có tính **trực giao (orthogonal) và bổ trợ cộng hưởng (synergistic)**. Tối ưu ngưỡng không thể thay thế năng lực trích xuất đặc trưng của P2, nhưng khi kết hợp sẽ tạo ra hiệu quả tổng thể vượt trội nhất.

---

#### 5. Trả Lời Trực Tiếp Câu Hỏi Nghiên Cứu RQ3
> **Câu hỏi nghiên cứu RQ3**: *Chiến lược tối ưu hóa ngưỡng quyết định nhạy cảm chi phí có thực sự giúp giảm tỷ lệ bỏ sót bo mạch phế phẩm (Board-level False Accept) và tỷ lệ lọt lỗi chí mạng (Critical False Negative) trong kiểm tra ngoại quan PCB hay không?*

- **Trả lời khẳng định: CÓ, RẤT RÕ RỆT VÀ MANG Ý NGHĨA THỰC TIỄN CAO.**
  1. **Về tỷ lệ lọt lỗi chí mạng ($Critical\ FN\ Rate$)**:
     - Trên kiến trúc chuẩn YOLOv8n: Giảm từ $23.87\%$ ($B_0$) xuống $21.75\%$ ($B_2$), tương ứng giảm tương đối $8.88\%$ nguy cơ.
     - Trên kiến trúc YOLOv8n-P2: Giảm từ $20.50\%$ ($B_1$) xuống còn $15.62\%$ ($Proposed$), tương ứng giảm tương đối $23.80\%$ nguy cơ.
     - Xét toàn diện từ Baseline gốc $B_0$ đến Giải pháp Đề xuất $Proposed$: $Critical\ FN\ Rate$ giảm từ **$23.87\%$ xuống $15.62\%$**, tương ứng **giảm tới $34.56\%$ nguy cơ bỏ sót các lỗi phá hủy linh kiện (ngắn mạch và hở mạch)**.
  2. **Về tỷ lệ lọt bo mạch phế phẩm ($Board\ FA$)**:
     - Tỷ lệ bỏ sót bo mạch lỗi được kéo giảm từ $0.63\%$ ($3/480$ bo mạch ở $B_0$) xuống còn **$0.42\%$** ($2/480$ bo mạch ở cả $B_1$, $B_2$ và $Proposed$).

---

#### 6. Giới Hạn Của Tập Dữ Liệu Nghiên Cứu (Limitation Statement)
- Toàn bộ 480 ảnh trong tập kiểm thử độc lập `data/splits/test.txt` của bộ dữ liệu `PKU-Market-PCB` đều là ảnh chụp bo mạch chứa khuyết tật ($100\%$ positive boards, 0 ảnh defect-free).
- **Hệ quả khoa học**: Chỉ số Tỷ lệ Báo động giả Cấp độ Bo mạch (**$Board\ FR$ - Board-level False Reject Rate**) không thể tính toán được một cách hợp lệ theo định nghĩa toán học chuẩn ($0 / 0 \rightarrow null/N/A$).
- Nghiên cứu ghi nhận rõ ràng đây là một hạn chế khách quan thuộc về bản chất cấu trúc phân phối của bộ dữ liệu benchmark quốc tế, không ảnh hưởng đến tính hợp lệ của chỉ số $Board\ FA$ cũng như các chỉ số $Recall$ và $Critical\ FN\ Rate$.

---

#### 7. Tuyên Bố Khóa Ngưỡng Kiểm Thử (Test Locked Statement)
- Tuân thủ nghiêm ngặt chuẩn mực liêm chính học thuật và phương pháp luận nghiên cứu thực nghiệm trong Thị giác Máy tính:
  - Tất cả các giá trị ngưỡng tin cậy $\tau_{B0} = 0.20$ và $\tau_{B1} = 0.13$ đều được lựa chọn duy nhất trên tập Validation và đã được khóa cố định trong tệp `configs/threshold.yaml` trước khi tiến hành kiểm thử.
  - **Tuyệt đối không thực hiện bất kỳ thao tác quét lưới (sweep), điều chỉnh tham số, hay tuning lại ngưỡng sau khi đã quan sát kết quả kiểm thử trên tập Test.**
  - Kết quả trên tập Test phản ánh chính xác và trung thực $100\%$ năng lực tổng quát hóa (generalization capability) trong điều kiện vận hành một lần duy nhất (one-shot locked evaluation) của mô hình đề xuất.


In [9]:
SUMMARY_MD_PATH = ROOT / "results" / "md" / "ablation_summary.md"
SUMMARY_MD_PATH.parent.mkdir(parents=True, exist_ok=True)

b0_crit = df_master.loc[df_master['code'] == 'B0', 'R_critical'].values[0]
b0_small = df_master.loc[df_master['code'] == 'B0', 'R_small'].values[0]
b0_fn = df_master.loc[df_master['code'] == 'B0', 'Critical_FN_rate'].values[0]

b1_crit = df_master.loc[df_master['code'] == 'B1', 'R_critical'].values[0]
b1_small = df_master.loc[df_master['code'] == 'B1', 'R_small'].values[0]
b1_fn = df_master.loc[df_master['code'] == 'B1', 'Critical_FN_rate'].values[0]

b2_crit = df_master.loc[df_master['code'] == 'B2', 'R_critical'].values[0]
b2_small = df_master.loc[df_master['code'] == 'B2', 'R_small'].values[0]
b2_fn = df_master.loc[df_master['code'] == 'B2', 'Critical_FN_rate'].values[0]

prop_crit = df_master.loc[df_master['code'] == 'Proposed', 'R_critical'].values[0]
prop_small = df_master.loc[df_master['code'] == 'Proposed', 'R_small'].values[0]
prop_fn = df_master.loc[df_master['code'] == 'Proposed', 'Critical_FN_rate'].values[0]

summary_lines = [
    "# TỔNG HỢP DIỄN GIẢI KẾT QUẢ THỰC NGHIỆM ABLATION (CHƯƠNG 4)",
    "",
    "## 1. Đóng góp của Kiến trúc FPN-P2 (B0 -> B1)",
    "- **Thiết kế thực nghiệm**: Cả hai cấu hình B0 và B1 đều áp dụng ngưỡng mặc định tau = 0.25.",
    "- **Phân tích kết quả**:",
    f"  - Nhánh phân giải cao P2 (160x160 pixels, stride-4) giúp tỷ lệ nhận diện khuyết tật kích thước nhỏ R_small tăng từ **{b0_small*100:.2f}% lên {b1_small*100:.2f}%** (+{(b1_small-b0_small)*100:.2f}%).",
    f"  - Tỷ lệ phát hiện khuyết tật chí mạng (open_circuit và short) R_critical tăng từ **{b0_crit*100:.2f}% lên {b1_crit*100:.2f}%** (+{(b1_crit-b0_crit)*100:.2f}%).",
    f"  - Tỷ lệ bỏ sót lỗi chí mạng (Critical_FN_rate) giảm từ **{b0_fn*100:.2f}% xuống {b1_fn*100:.2f}%** (giảm tương đối 14.12% rủi ro lọt lỗi).",
    "  - Tỷ lệ bỏ lọt bo mạch phế phẩm (Board_FA) giảm từ **0.63% xuống 0.42%** (chỉ còn 2/480 bo mạch bị bỏ lọt).",
    "- **Ý nghĩa kỹ thuật**: Nhánh P2 duy trì độ phân giải không gian lớn gấp 4 lần nhánh P3, bảo toàn hữu hiệu các đặc trưng cục bộ của vết hở mạch và ngắn mạch vi mô qua các phép pooling/conv sâu.",
    "",
    "---",
    "",
    "## 2. Đóng góp của Tối Ưu Hóa Ngưỡng Nhạy Cảm Chi Phí (B0 -> B2)",
    "- **Thiết kế thực nghiệm**: Giữ nguyên mô hình YOLOv8n chuẩn, thay đổi ngưỡng mặc định tau = 0.25 sang ngưỡng tối ưu F2-Max đã khóa trên validation tau_B0 = 0.20.",
    "- **Phân tích kết quả**:",
    f"  - Không cần sửa đổi kiến trúc hay tốn thêm chi phí tính toán, việc hạ ngưỡng tối ưu giúp R_small tăng từ **{b0_small*100:.2f}% lên {b2_small*100:.2f}%** (+{(b2_small-b0_small)*100:.2f}%).",
    f"  - R_critical tăng từ **{b0_crit*100:.2f}% lên {b2_crit*100:.2f}%** (+{(b2_crit-b0_crit)*100:.2f}%).",
    f"  - Critical_FN_rate giảm từ **{b0_fn*100:.2f}% xuống {b2_fn*100:.2f}%** (giảm tương đối 8.88% nguy cơ bỏ sót lỗi chí mạng).",
    "  - Board_FA giảm từ **0.63% xuống 0.42%**.",
    "- **Ý nghĩa kỹ thuật**: Tối ưu ngưỡng theo mục tiêu F2 (beta = 2) dịch chuyển điểm cân bằng về phía ưu tiên Recall, vớt lại các khuyết tật có độ tin cậy thấp do nhiễu nền hoặc ánh sáng yếu.",
    "",
    "---",
    "",
    "## 3. Tác Động Khi Tích Hợp Ngưỡng Tối Ưu Vào Kiến Trúc P2 (B1 -> Proposed)",
    "- **Thiết kế thực nghiệm**: Áp dụng ngưỡng tối ưu F2-Max tau_B1 = 0.13 (đã khóa trên validation) cho mô hình YOLOv8n-P2 thay vì dùng ngưỡng mặc định 0.25.",
    "- **Phân tích kết quả**:",
    "  - Đạt bước đột phá lớn nhất trong toàn bộ ma trận ablation:",
    f"    - R_small tăng vọt từ **{b1_small*100:.2f}% lên {prop_small*100:.2f}%** (+{(prop_small-b1_small)*100:.2f}%).",
    f"    - R_critical tăng vọt từ **{b1_crit*100:.2f}% lên {prop_crit*100:.2f}%** (+{(prop_crit-b1_crit)*100:.2f}%).",
    f"    - Critical_FN_rate bị cắt giảm sâu từ **{b1_fn*100:.2f}% xuống chỉ còn {prop_fn*100:.2f}%** (giảm tương đối tới 23.80% rủi ro lọt lỗi so với chính B1).",
    "    - Board_FA duy trì ổn định ở mức tối ưu **0.42%**.",
    "- **Ý nghĩa kỹ thuật**: Tầng P2 tạo ra rất nhiều đặc trưng vi mô chuẩn xác nhưng có phân phối xác suất nằm trong dải [0.13, 0.25). Việc hạ ngưỡng xuống tau = 0.13 đã khai phóng toàn bộ tiềm năng phát hiện khuyết tật tiềm ẩn của kiến trúc P2.",
    "",
    "---",
    "",
    "## 4. So Sánh Độc Lập Khi Cùng Dùng Ngưỡng Tối Ưu (B2 vs Proposed)",
    "- **Thiết kế thực nghiệm**: So sánh B2 (YOLOv8n, tau = 0.20) và Proposed (YOLOv8n-P2, tau = 0.13) khi cả hai đều đã đạt trạng thái tối ưu ngưỡng của mình.",
    "- **Phân tích kết quả**:",
    "  - Ngay cả khi baseline đã được tối ưu ngưỡng tối đa (B2), kiến trúc P2 trong Proposed vẫn mang lại bước tiến vượt bậc:",
    f"    - R_small: Proposed ({prop_small*100:.2f}%) vượt trội hơn B2 ({b2_small*100:.2f}%) là **+{(prop_small-b2_small)*100:.2f}%**.",
    f"    - R_critical: Proposed ({prop_crit*100:.2f}%) vượt trội hơn B2 ({b2_crit*100:.2f}%) là **+{(prop_crit-b2_crit)*100:.2f}%**.",
    f"    - Critical_FN_rate: Proposed ({prop_fn*100:.2f}%) thấp hơn B2 ({b2_fn*100:.2f}%) tới **{(b2_fn-prop_fn)*100:.2f}%** điểm phần trăm (giảm tương đối 28.18% nguy cơ lọt lỗi so với B2).",
    "- **Ý nghĩa khoa học**: Cải tiến kiến trúc P2 và tối ưu hóa ngưỡng chi phí F2 là hai giải pháp mang tính **trực giao (orthogonal) và bổ trợ cộng hưởng (synergistic)**. Tối ưu ngưỡng không thể thay thế biểu diễn đặc trưng phân giải cao của P2, nhưng khi kết hợp sẽ tạo nên hiệu quả tối đa.",
    "",
    "---",
    "",
    "## 5. Trả Lời Trực Tiếp Câu Hỏi Nghiên Cứu RQ3",
    "> **Câu hỏi nghiên cứu RQ3**: Chiến lược tối ưu hóa ngưỡng quyết định nhạy cảm chi phí có thực sự giúp giảm tỷ lệ bỏ sót bo mạch phế phẩm (Board-level False Accept) và tỷ lệ lọt lỗi chí mạng (Critical False Negative) trong kiểm tra ngoại quan PCB hay không?",
    "",
    "- **Khẳng định: CÓ, RẤT RÕ RỆT VÀ MANG LẠI Ý NGHĨA THỰC TIỄN CAO TRONG CÔNG NGHIỆP.**",
    "  1. **Về Critical False Negative Rate**:",
    f"     - B0 -> B2: Giảm từ {b0_fn*100:.2f}% xuống {b2_fn*100:.2f}% (giảm tương đối 8.88% lỗi chí mạng).",
    f"     - B1 -> Proposed: Giảm từ {b1_fn*100:.2f}% xuống {prop_fn*100:.2f}% (giảm tương đối 23.80% lỗi chí mạng).",
    f"     - Toàn diện B0 -> Proposed: Giảm từ **{b0_fn*100:.2f}% xuống {prop_fn*100:.2f}%** (cắt giảm tới **34.56%** nguy cơ lọt lỗi nghiêm trọng hở mạch và ngắn mạch).",
    "  2. **Về Board-level False Accept Rate (Board_FA)**:",
    "     - Giảm từ 0.63% (3/480 bo mạch ở B0) xuống còn **0.42%** (2/480 bo mạch ở Proposed).",
    "",
    "---",
    "",
    "## 6. Giới Hạn Của Tập Dữ Liệu Nghiên Cứu (Limitation Statement)",
    "- Toàn bộ 480 ảnh trong tập test của PKU-Market-PCB đều chứa khuyết tật (100% defective boards, không có ảnh PASS/defect-free).",
    "- **Hệ quả**: Chỉ số Tỷ lệ Báo động giả cấp độ Bo mạch (**Board_FR - Board-level False Reject Rate**) không thể đánh giá được theo đúng định nghĩa toán học (mẫu số bằng 0 -> null/N/A).",
    "- Đây là giới hạn khách quan của bộ dữ liệu benchmark công khai, không ảnh hưởng đến tính toàn vẹn của các chỉ số Recall, Critical FN Rate và Board FA.",
    "",
    "---",
    "",
    "## 7. Tuyên Bố Khóa Ngưỡng Kiểm Thử (Test Locked Statement)",
    "- Tuân thủ liêm chính học thuật và phương pháp luận nghiên cứu thực nghiệm:",
    "  - Các ngưỡng tau_B0 = 0.20 và tau_B1 = 0.13 được chọn duy nhất trên tập Validation và đã khóa trong configs/threshold.yaml trước khi chạy test.",
    "  - **Tuyệt đối không quét lại ngưỡng, không chọn lại hay tinh chỉnh ngưỡng sau khi quan sát kết quả trên tập Test.**",
    "  - Đánh giá trên tập Test là kiểm định một lần duy nhất (one-shot locked evaluation), bảo đảm tính khách quan và khả năng tổng quát hóa thực tế của đề tài.",
]

summary_md_content = "\n".join(summary_lines) + "\n"

with open(SUMMARY_MD_PATH, "w", encoding="utf-8") as fp:
    fp.write(summary_md_content)

print("=" * 80)
print(f"✅ ĐÃ XUẤT THÀNH CÔNG BẢN TÓM TẮT DIỄN GIẢI CHƯƠNG 4 TẠI:")
print(f"   {SUMMARY_MD_PATH.relative_to(ROOT)}")
print("=" * 80)


✅ ĐÃ XUẤT THÀNH CÔNG BẢN TÓM TẮT DIỄN GIẢI CHƯƠNG 4 TẠI:
   results/md/ablation_summary.md


In [10]:
summary_deltas = []
for _, r in df_master.iterrows():
    c = r['code']
    d_small = (r['R_small'] - b0_small) * 100
    d_crit = (r['R_critical'] - b0_crit) * 100
    d_fn = (r['Critical_FN_rate'] - b0_fn) * 100
    rel_fn_reduct = ((b0_fn - r['Critical_FN_rate']) / b0_fn) * 100
    summary_deltas.append({
        "Cấu hình": c,
        "R_small (%)": f"{r['R_small']*100:.2f}%",
        "Δ R_small": f"{d_small:+.2f}%",
        "R_critical (%)": f"{r['R_critical']*100:.2f}%",
        "Δ R_critical": f"{d_crit:+.2f}%",
        "Critical FN Rate": f"{r['Critical_FN_rate']*100:.2f}%",
        "Giảm rủi ro FN tương đối": f"{rel_fn_reduct:.2f}%" if c != "B0" else "-",
    })

df_deltas = pd.DataFrame(summary_deltas)
print("=" * 80)
print("BẢNG TỔNG HỢP MỨC ĐỘ CẢI THIỆN TƯƠNG ĐỐI SO VỚI BASELINE B0:")
print("=" * 80)
print(df_deltas.to_string(index=False))
print("=" * 80)


BẢNG TỔNG HỢP MỨC ĐỘ CẢI THIỆN TƯƠNG ĐỐI SO VỚI BASELINE B0:
Cấu hình R_small (%) Δ R_small R_critical (%) Δ R_critical Critical FN Rate Giảm rủi ro FN tương đối
      B0      68.21%    +0.00%         76.12%       +0.00%           23.87%                        -
      B1      69.50%    +1.29%         79.50%       +3.38%           20.50%                   14.12%
      B2      70.08%    +1.87%         78.25%       +2.13%           21.75%                    8.88%
Proposed      74.58%    +6.37%         84.38%       +8.26%           15.62%                   34.56%
